File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        def is_sign(char) -> bool:
            sign = ['+', '-', '*', '/']
            if char in sign:
                return True
            return False

        def cal(n1, n2, op):
            if op == '+':
                return n1+n2
            elif op == '-':
                return n1-n2
            elif op == '*':
                return n1*n2
            else:
                return int(n1/n2)
        
        for s in tokens:
            if is_sign(s) and len(stack) >= 2:
                n1 = stack.pop()
                n2 = stack.pop()
                ret = cal(n2, n1, s)
                stack.append(ret)
            else:
                stack.append(int(s))
        
        return stack.pop()

File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.cpp
#include <algorithm>

class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> st;
        for (auto s: tokens) {
            if (isOperation(s)) {
                int n1 = st.top();
                st.pop();
                int n2 = st.top();
                st.pop();
                int ret = calculate(n2, n1, s[0]);
                st.push(ret);
            } else {
                st.push(stoi(s));
            }
        }

        return st.top();
    }

    bool isOperation(string s) {
        if (s.size() != 1) return false;
        array<char, 4> operations{{'+', '-', '*', '/'}};
        
        auto ptr = find(operations.begin(), operations.end(), s[0]);
        return ptr != operations.end();
    }

    int calculate(int n1, int n2, char op) {
        if (op == '+') {
            return n1+n2;
        } else if (op == '-') {
            return n1-n2;
        } else if (op == '*') {
            return n1*n2;
        } else {
            return (int)(n1/n2);
        }
    }
};